### DISCUSSION TITLE
data from PDF documents in Lua filters?

### DISCUSSION THREAD

**Discussion by pr-apes** [speedata/publisher#388] [Q&A] [answered]
@pgundlach,

I would like to rewrite for Publisher some basic hacks that were written in ConTeXt.

These scripts play with page imposition and image clipping.

I guess that the filters allow standard Lua code and some modules (it would be great to know which ones are available).

But besides access to variables (option `-v`), I wonder whether it is possible to have access in the filter to the following PDF info:

1. Dimensions: height, width, number of pages.

1. Metadata: author, title, subject...

Many thanks for your help.

**Comment by pgundlach** [comment, marked as answer]
No. You can get the number of pages by reading the `-aux.xml` file and analyse that. 

What you can do is write the data you need to an external XML file ([SaveDataset](https://doc.speedata.de/publisher/en/commandreference/savedataset/)) and parse the data after the run.

**Reply by pr-apes** [reply]
Many thanks for your reply, @pgundlach.

Sorry, but I’m not sure I get the point (probably because I may be aiming at something not achievable by _Publisher_).

I have a bunch of PDF documents with different page sizes and total numbers of pages (they are generated by other tools).

ConTeXt runs a kind of _Lua_ script (with “PDF superpowers”) that depending on the page dimensions makes a different imposition, such as the following ones:

* If page size is landscape legal, please clip it to landscape A4.
* If page size is portrait legal, please append it to another portrait A4, after clipping it to A4 page size.
* If page size is portrait A5, please clip a predefined image area and impose it centered (both vertically and horizontally) in an A4 page.
* And some other cases...

_Lua_ filters cannot read dimensions or other data from external PDF documents.

To be able to read these dimensions, the following code will not be of use:

``` xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Record element="data">
    <SetVariable variable="myfile" select="$external_pdfdoc"/>
    <Loop select="sd:number-of-pages($myfile)" variable="page">
      <PlaceObject column="0mm" row="0mm">
        <Image file="{$myfile}" width="210mm" page="{$page}"/>
      </PlaceObject>
      <ClearPage/>
    </Loop>
  </Record>

</Layout>
```

Because the image is placed on an already given pages, that will be reading page size dimensions of the main PDF documents.

I wonder whether there is something similar in _Publisher_ to what a TEXpage is in ConTeXt (page with no extra dimensions that the ones required by its contents, such as an image).

But I see that I can have that with XPath functions (no metadata, though): `sd:aspectratio()`, `sd:imageheight()`, `sd:imagewidth()` (and `sd:aspectratio()`).

This could help to do it:

``` xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Record element="data">
    <SetVariable variable="myfile" select="$external_pdfdoc"/>
	<Output>
	<Text>
	<Paragraph>
    <Value select="sd:number-of-pages($myfile)"/>
	<Value> </Value>
    <Value select="sd:imageheight($myfile)"/>
	<Value> </Value>
    <Value select="sd:imagewidth($myfile)"/>
	</Paragraph>
      </Text>
    </Output>
  </Record>

</Layout>
```

The only minor issue is that the measurement unit is grid cells.

I think this would set grid cells in millimiters:

``` xml
<SetGrid width="1mm" height="1mm"/>
```

Now I see how it might work.

Many thanks for your help.

**Reply by pgundlach** [reply]
I didn't think of external PDFs. The data from these files are not accessible in the preprocessing step.

Imageheight and imagewitdth have an undocumented feature (sorry - will be documented in the next version):

`sd:imageheight($myfile, 'mm')` returns the number of millimetres (same with `imagewidth()`).

**Reply by pr-apes** [reply]
That would be a great feature.

**Reply by pr-apes** [reply]
Even greater would be `sd:pdfauthor()` `sd:title()` and similar ones...

**Comment by pr-apes** [comment]
Sorry, I get it know, it was not an unimplemented feature, just an undocumented one.
